cabal-install/Distribution/Client/ProjectBuilding.hs

```diff
@@ -349,22 +351,23 @@ improveInstallPlanWithUpToDatePackages :: ElaboratedInstallPlan
                                        -> ElaboratedInstallPlan
 improveInstallPlanWithUpToDatePackages installPlan pkgsBuildStatus =
     replaceWithPrePreExisting installPlan
-      [ (installedPackageId pkg, ipkgs)
+      [ (installedPackageId pkg, mipkg)
       | InstallPlan.Configured pkg
           <- InstallPlan.reverseTopologicalOrder installPlan
       , let ipkgid = installedPackageId pkg
             Just pkgBuildStatus = Map.lookup ipkgid pkgsBuildStatus
-      , BuildStatusUpToDate (BuildResult { buildResultLibInfo = ipkgs })
+      , BuildStatusUpToDate (BuildResult { buildResultLibInfo = mipkg })
           <- [pkgBuildStatus]
       ]
   where
     replaceWithPrePreExisting =
-      foldl' (\plan (ipkgid, ipkgs) ->
-                case find (\ipkg -> installedPackageId ipkg == ipkgid) ipkgs of
+      foldl' (\plan (ipkgid, mipkg) ->
+                case mipkg of
                   Just ipkg -> InstallPlan.preexisting ipkgid ipkg plan
-                  Nothing   -> unexpected)
-    unexpected =
-      error "improveInstallPlanWithUpToDatePackages: dep on non lib package"
+                  -- TODO: Maybe this is a little wrong, because
+                  -- pre-installed executables show up in the
+                  -- InstallPlan as source packages.
+                  Nothing -> plan)
```

Comment:
I think I'll follow up with a patch to add a new constructor for Installed, distinct from PreExisting, containing a srcpkg (rather than ipkg) and perhaps the lib ipkg too if it's actually needed.